base/compiler/ssair/irinterp.jl

```diff
@@ -225,16 +225,21 @@ function reprocess_instruction!(interp::AbstractInterpreter,
             (; rt, effects) = abstract_eval_statement_expr(interp, inst, nothing, ir, irsv.mi)
             # All other effects already guaranteed effect free by construction
             if is_nothrow(effects)
+                ir.stmts[idx][:flag] |= IR_FLAG_EFFECT_FREE | IR_FLAG_NOTHROW
```

Comment:
`IR_FLAG_EFFECT_FREE` seems a bit risky here since we may prove `:effect_free`-ness in a presence of effect-ful statement, e.g.
```julia
julia> mutable struct SafeRef{T}
           x::T
       end

julia> Base.getindex(x::SafeRef) = x.x

julia> Base.setindex!(x::SafeRef, y) = x.x = y

julia> Base.infer_effects((String,String,)) do x, y
           r = SafeRef(x)
           r[] = y
           r[]
       end
(+c,+e,+n,+t,+s,+m)
```

So I guess we need to check `is_removable_if_unused` here?

Reply:
I think you're probably right, though IR_FLAG_EFFECT_FREE was there before, so this might be an existing bug. Ideally we should have a testcase for this.

Reply:
I attempted to create a test case to trigger an error because of this issue, but it seems very difficult to invent one. Something like the example below would be close (it actually crashes due to a different bug in our SROA though).
```julia
Base.@assume_effects :consistent :nothrow @inline function foo(a::Int, b)
    r = Ref(a)
    try
        setfield!(r, :x, b)
        nothing
    catch err
        return isa(b, String)
    end
end
```

Reply:
xref: #48067